### DISCUSSION TITLE
How to change specific values of previously materialized asset?

### DISCUSSION THREAD

**Discussion by ammerzon** [dagster-io/dagster#15149] [Q&A] [answered]
**Context**

I am using Dagster to fetch data from an API weekly. For instance, a list of published books where the response would look something like this:

```json
[
  {
    "title": "title",
    "published_at": "2021-08-04T13:00:00.000",
    "modified_at": "2021-10-04T13:00:00.000",
    "description": "description"
  },
  ...
]
```

**Problem**

My approach is to partition the data by week but sometimes, the data of the API changes. For instance, the description is adapted and changes the `modified_at` field. Fortunately, the API offers a way to query based on the lastModStartDate and lastModEndDate. My question now is whether I can change only certain data of an already materialized asset and, if not, how I should tackle this problem.

**Comment by jamiedemaria** [comment]
Hey @ammerzon - can you clarify a couple things about your use case? 

From what I understand you will have a partitioned asset where each partition is a week of data from the published books API. 

Since the records can be updated the issue is that a record from 2 weeks ago could be modified. I'm assuming you wouldn't want to re-materialize the entire partition just to update that one record, so instead you'd like to go in and just update the single record.

is that all correct?

**Reply by ammerzon** [reply]
Yes, @jamiedemaria that‘s exactly what I want to achieve. I want to avoid to re-materialize the whole partition if just a single record needs to be changed 😁

**Comment by sryza** [comment]
Hey @ammerzon how attached are you to using partitions. One recommendation, instead of using partitions, just update your asset incrementally: https://github.com/dagster-io/dagster/discussions/14733.

**Reply by ammerzon** [reply]
I quite like them because it offers me the flexibility to:
- rerun failed materializations where I don't have to materialize the whole asset again 
- use IOManagers

This wouldn't be possible with your recommendation @sryza, right?

But I am open the other solutions if my use case isn't possible with partitions.

**Reply by ammerzon** [reply]
@sryza, is it possible to access the last materialized state of an asset in the asset itself?
Because then I could reuse values and avoid recomputations also in a partition.

**Reply by sryza** [reply]
Hey - take a look at this: https://github.com/dagster-io/dagster/discussions/14805

**Reply by ammerzon** [reply, marked as answer]
I think I'll then stick with partitions and create an asset that works like this:

```python
@asset(partitions_def=WeeklyPartitionsDefinition(start_date="2022-01-01"))
def books(context):
    from module_that_contains_defs import defs

    fetched_books = fetch_changed_books()
    books_prior_value = defs.load_asset_value("books")
    
    updated_books = []
    for book in fetched_books:
        prior_book = next((item for item in books_prior_value if item["id"] == book["id"]), None)
        if prior_book and book['updated_at'] != prior_book['updated_at']:
            # do computation here
            updated_books.append(book)
        else:
            updated_books.append(prior_book)
            
    return updated_books
```

Thank you for your help @sryza 😁
